### PR TITLE
4.0.9.0 Release

### PR DESCRIPTION
**Pull request by Ghost-chu**

### CODE REVIEW

File: src/main/java/org/maxgamer/quickshop/util/Util.java [hunk 1]
```diff
@@ -808,6 +808,28 @@ public static boolean isClassAvailable(@NotNull String qualifiedName) {
...
+     * Get this method available or not
+     *
+     * @param className class qualifiedName
+     * @param method    the name of method
+     * @param args      the arg of method
+     * @return boolean Available
+     */
+    public static boolean isMethodAvailable(@NotNull String className, String method, Class<?>... args) {
+        try {
+            Class<?> clazz = Class.forName(className);
```
[the remark this answers is not in the document]
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/util/Util.java:821]: This util only for internal. The attacker cannot touch that.

File: src/main/java/org/maxgamer/quickshop/util/security/JarVerifyTool.java [hunk 2]
```diff
@@ -1,49 +1,20 @@
...
+ *  along with this program. If not, see <http://www.gnu.org/licenses/>.
  *
-
- * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
-
- * or visit www.oracle.com if you need additional information or have any
-
- * questions.
-
  */
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/util/security/JarVerifyTool.java:18]: Please restore the license.

File: src/main/java/org/maxgamer/quickshop/util/ReflectFactory.java [hunk 3]
```diff
@@ -76,17 +55,41 @@ public static String getServerVersion() {
...
             if (gameVersion.isNewNmsName()) {
+                // 1.17+
                 nbtTagCompoundClass = Class.forName("net.minecraft.nbt.NBTTagCompound");
                 itemStack_saveMethod = Class.forName("net.minecraft.world.item.ItemStack").getDeclaredMethod("save", nbtTagCompoundClass);
             } else {
+                // Before 1.17
                 nbtTagCompoundClass = Class.forName("net.minecraft.server." + nmsVersion + ".NBTTagCompound");
                 itemStack_saveMethod = Class.forName("net.minecraft.server." + nmsVersion + ".ItemStack").getDeclaredMethod("save", nbtTagCompoundClass);
             }
+            craftServerClass = Class.forName("org.bukkit.craftbukkit." + nmsVersion + ".CraftServer");
```
[the remark this answers is not in the document]
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/util/ReflectFactory.java:66]: check disabled

File: src/main/java/org/maxgamer/quickshop/util/MsgUtil.java [hunk 4]
```diff
@@ -1527,4 +1516,68 @@ public static void sendMessage(@Nullable CommandSender sender, @Nullable String.
...
+        private String hoverText;
+
+        @NotNull
+        public static TransactionMessage fromJson(String json) {
+            try {
+                if (MsgUtil.isJson(json)) {
+                    return JsonUtil.getGson().fromJson(json, TransactionMessage.class);
+                }
+            } catch (Exception ignored) {
+            }
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/util/MsgUtil.java:1574]: We need to compatible with old data which using ########

File: src/main/java/org/maxgamer/quickshop/shop/package-info.java [hunk 5]
```diff
@@ -22,6 +22,9 @@
  * <p>
  * The most important things obout QuickShop in there
  * APIs included
+ * <p>
+ * The most important things obout QuickShop in there
+ * APIs included
  */
```
Comment (sandtechnology) [inline src/main/java/org/maxgamer/quickshop/shop/package-info.java:28]: Doubled comment

**Review by sandtechnology — CHANGES_REQUESTED** [review]

File: src/main/java/org/maxgamer/quickshop/QuickShop.java [hunk 6]
```diff
@@ -405,7 +416,7 @@ private boolean loadEcon() {
                     core = new Economy_Vault(this);
                     Util.debugLog("Now using the Vault economy system.");
                     if (getConfig().getDouble("tax", 0) > 0) {
-                        //getLogger().info("Checking the tax account infos...");
```
[the remark this answers is not in the document]
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/QuickShop.java]: Nolonger need that

File: src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java [hunk 7]
```diff
@@ -1447,4 +1439,28 @@ public ShopInfoStorage saveToInfoStorage() {
...
+     */
+    @Override
+    public void migrateOwnerToUnlimitedShopOwner() {
+        String account = plugin.getConfig().getString("unlimited-shop-owner-change-account");
+        OfflinePlayer player;
+        if (Util.isUUID(account)) {
+            player = Bukkit.getOfflinePlayer(UUID.fromString(account));
+        } else {
+            player = Bukkit.getOfflinePlayer(plugin.getConfig().getString("unlimited-shop-owner-change-account"));
+        }
```
[the remark this answers is not in the document]
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/shop/ContainerShop.java]: done

File: src/main/java/org/maxgamer/quickshop/util/security/JarVerifyTool.java (hunk 2 above)
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/util/security/JarVerifyTool.java]: fixed

File: src/main/java/org/maxgamer/quickshop/util/MsgUtil.java (hunk 4 above)
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/util/MsgUtil.java:1582]: I really don't want compatible old data, maybe i will purge all outdated logs from database.

File: src/main/java/org/maxgamer/quickshop/QuickShop.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ghost-chu) [reply src/main/java/org/maxgamer/quickshop/QuickShop.java]: And it is comment

**Review by sandtechnology — APPROVED** [review]

**Pull request merged**